README.md
rackspace_memcached Cookbook
==================

Installs memcached and provides a template for its main configuration file.

Requirements
------------

### Platforms
- CentOS 6.4
- Ubuntu 12.04
- Debian 7.2

May work on other systems with or without modification.

### Cookbooks
This cook book does not depend on any other cookbooks

Attributes
----------
The following are node attributes passed to the template:

- `['rackspace_memcached']['config']['-m']['value']`` - maximum memory for memcached instances, in MB. Default is 64.
- `['rackspace_memcached']['config']['-u']['value']`` - user to run memcached as.
- `['rackspace_memcached']['config']['-p']['value']`` - TCP port for memcached to listen on. Default is 11211.
- `['rackspace_memcached']['config']['-U']['value']`` - UDP port for memcached to listen on. Default is 11211.
- `['rackspace_memcached']['config']['-l']['value']`` - IP address for memcache to listen on, defaults to **0.0.0.0** (world accessible).
- `['rackspace_memcached']['config']['-c']['value']`` - maximum number of connections to accept. Defaults is 1024.
- `['rackspace_memcached']['config']['-I']['value']`` - maximum size of an object to cache. Default is 1MB.
- `['rackspace_memcached']['config']['logfile']['value']` - logfile to which memcached output will be redirected. Default is /var/log/memcached.log.
- `['rackspace_memcached']['include_dev_lib']` - whether or not to install development libraries as well. Default is true


Usage
-----
Simply set the attributes and the cookbook will configure the `/etc/memcached.conf` file (Ubuntu, Debian) or /etc/sysconfig/memcached file (CentOS, RHEL). If you want to use multiple memcached instances, you'll need to modify the recipe to disable the startup script and the template in the default recipe.

To add an option to the /etc/memcached.conf file (Ubuntu/Debian) that does not have a default value from this cookbook and is not enabled/set by default in memcache, simply define a new attribute that follows the pattern of those in the Attributes section above. For example, to enable extra verbosity, add the following line to your recipe:

`node.default['rackspace_memcached']['config']['-vv']['value'] = true`

The template will render a line with content `-vv`. Note that for this example, you must set the value to boolean `true`, *not* a string `'true'`.

To add an option to the configuration that takes a value (e.g. `-m 128`), you must define the attribute like the following:

`node.default['rackspace_memcached']['config']['-m']['value'] = 128`


Contributing
------------
* See contributing guidelines [here](https://github.com/rackspace-cookbooks/contributing/blob/master/CONTRIBUTING.md)

Testing
-------
* See contributing guidelines [here](https://github.com/rackspace-cookbooks/contributing/blob/master/CONTRIBUTING.md)

License & Authors
-----------------
- Author:: Joshua Timberman (<joshua@opscode.com>)
- Author:: Joshua Sierles (<joshua@37signals.com>)
- Author:: Kent Shultz (<kent.shultz@rackspace.com>)

```text
Copyright:: 2009-2012, Opscode, Inc
Copyright:: 2009, 37signals
Copyright:: 2014, Rackspace, US Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
rackspace_memcached Cookbook CHANGELOG
============================
This file is used to list changes made in each version of the memcached cookbook.

V2.0.4
------
- Add attribute to decide if we should install dev libraries for memcache
- remove strainer
- chefspec coverage

v2.0.3
------
- change logrotate template to logrotate_app definition
  - this adds a dependency on the rackspace_logrotate cookbook
- generalize the template for /etc/memcached.conf to walk the config hash
  - supports value/comment convention in the config hash data structure

v2.0.2
------
- Add ['config'] into config hash
- Add logrotate
- Make logfilename attribute fully qualified

v2.0.1
------
- now run memcached as 'memcache' user on ubuntu, not 'nobody' 

v2.0.0
------
Rackspace rebuild changes
- removed support for unsupported OSes (namely: smartos, fedora, suse)
- removed dependency on yum; it was only necessary before due to now-deprecated RHEL 5 support
- removed dependency on runit and definition memcached_instance


v1.7.0
------
Updating for yum ~> 3.0. 
Fixing up style issues for rubocop. 
Updating test-kitchen harness


v1.6.6
------
fixing metadata version error. locking to 3.0


v1.6.4
------
Locking yum dependency to '< 3'


v1.6.2
------
[COOK-3741] UDP settings for memcached


v1.6.0
------
### Bug
- **[COOK-3682](https://tickets.opscode.com/browse/COOK-3682)** - Set user when using Debian packages

### Improvement
- **[COOK-3336](https://tickets.opscode.com/browse/COOK-3336)** - Add an option to specify the logfile (fix)

v1.5.0
------
### Improvement
- **[COOK-3336](https://tickets.opscode.com/browse/COOK-3336)** - Add option to specify logfile
- **[COOK-3299](https://tickets.opscode.com/browse/COOK-3299)** - Document that `memcached` is exposed by default

### Bug
- **[COOK-2990](https://tickets.opscode.com/browse/COOK-2990)** - Include `listen`, `maxconn`, and `user` in the runit service

### New Feature
- **[COOK-2790](https://tickets.opscode.com/browse/COOK-2790)** - Add support for defining max object size

v1.4.0
------
### Improvement
- [COOK-2756]: add SUSE support to memcached cookbook
- [COOK-2791]: Remove the template for Karmic from the memcached cookbook

### Bug
- [COOK-2600]: support memcached on SmartOS

v1.3.0
------
- [COOK-2386] - update `memcached_instance` definition for `runit_service` resource

v1.2.0
------
- [COOK-1469] - include yum epel recipe on RHEL 5 (introduces yum cookbook dependency)
- [COOK-2202] - Fix typo in previous ticket/commits
- [COOK-2266] - pin runit dependency

v1.1.2
------
- [COOK-990] - params insite runit_service isn't the same as outside

v1.1.0
------
- [COOK-1764] - Add Max Connections to memcached.conf and fix typos

v1.0.4
------
- [COOK-1192] - metadata doesn't include RH platforms (supported)
- [COOK-1354] - dev package changed name on centos6

v1.0.2
------
- [COOK-1081] - support for centos/rhel

v1.0.0
------
- [COOK-706] - Additional info in README
- [COOK-828] - Package for RHEL systems

v0.10.4
-------
- Current released version


spec/unit/recipes/default_spec.rb
require 'spec_helper'

describe 'rackspace_memcached::default' do
  let(:chef_run) { ChefSpec::Runner.new.converge(described_recipe) }

  it 'installs the memcached package' do
    expect(chef_run).to install_package('memcached')
  end

  it 'creates memcached service with :nothing action' do
    expect(chef_run).to_not start_service('memcached')
    expect(chef_run).to_not enable_service('memcached')
  end

  context 'on rhel' do
    let(:chef_run) { ChefSpec::Runner.new(platform: 'redhat', version: '6.4').converge(described_recipe) }
    let(:template) { chef_run.template('/etc/sysconfig/memcached') }

    it 'installs libmemcached-devel package'do
      expect(chef_run).to install_package('libmemcached-devel')
    end

    it 'writes the /etc/sysconfig/memcached' do
      expect(chef_run).to create_template('/etc/sysconfig/memcached').with(
        user: 'root',
        group: 'root',
        mode: '0644'
      )
    end

    it 'notifies the service to restart' do
      expect(template).to notify('service[memcached]').to(:restart)
    end
  end

  context 'on ubuntu' do
    let(:chef_run) { ChefSpec::Runner.new(platform: 'ubuntu', version: '12.04').converge(described_recipe) }
    let(:template) { chef_run.template('/etc/memcached.conf') }

    it 'installs libmemcache-dev package'do
      expect(chef_run).to install_package('libmemcache-dev')
    end

    it 'writes the /etc/memcached.conf' do
      expect(chef_run).to create_template('/etc/memcached.conf').with(
        user: 'root',
        group: 'root',
        mode: '0644'
      )
    end

    it 'notifies the service to restart' do
      expect(template).to notify('service[memcached]').to(:restart)
    end
  end

end


test/fixtures/cookbooks/fake/metadata.rb
name    'fake'
version '1.0.0'

depends 'rackspace_memcached'


test/fixtures/cookbooks/with_maas/metadata.rb
name    'with_maas'
version '1.0.0'

depends 'rackspace_memcached'
depends 'rackspace_cloudmonitoring'


recipes/monitors.rb
#
# Cookbook Name:: rackspace_memcached
# Recipe:: default
#
# Copyright 2014, Rackspace, US Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

cookbook_file '/usr/lib/rackspace-monitoring-agent/plugins/memcached_stats.py' do
  source  'memcached_stats.py'
  owner   'root'
  group   'root'
  mode    '0755'
  only_if 'test -d /usr/lib/rackspace-monitoring-agent/plugins'
end

# optionally specify alarm and notification_plan on a rolebook level
node.default['rackspace_cloudmonitoring']['monitors']['memcached_stats']['type'] = 'agent.plugin'
node.default['rackspace_cloudmonitoring']['monitors']['memcached_stats']['details'] = { file: 'memcached_stats.py',
                                                                                        args: ['localhost', node['rackspace_memcached']['config']['-p']['value']]
                                                                                      }


attributes/default.rb
#
# Cookbook Name:: rackspace_memcached
# Attributes:: default
#
# Copyright 2014, Rackspace, US Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
default['rackspace_memcached']['templates_cookbook']['sysconfig'] = 'rackspace_memcached'
default['rackspace_memcached']['templates_cookbook']['memcached_conf'] = 'rackspace_memcached'
default['rackspace_memcached']['config']['-d'] = { value: true, comment: 'run as daemon' }
default['rackspace_memcached']['config']['logfile'] = { value: '/var/log/memcached.log' }
default['rackspace_memcached']['config']['-v'] = { value: true, comment: 'verbose output' }
default['rackspace_memcached']['config']['-m'] = { value: 64, comment: 'cache size' }
default['rackspace_memcached']['config']['-p'] = { value: '11211', comment: 'port' }
default['rackspace_memcached']['config']['-U'] = { value: '11211', comment: 'udp port' }
default['rackspace_memcached']['config']['-l'] = { value: '0.0.0.0', comment: 'listen address' }
default['rackspace_memcached']['config']['-c'] = { value: 1024, comment: 'maximum incoming connections' }
default['rackspace_memcached']['config']['-I'] = { value: '1m', comment: 'maximum object size' }
default['rackspace_memcached']['config']['-u']['comment'] = 'run as user instead of root'
# default['rackspace_memcached']['config']['-vv'] = { value: true, comment: 'extra verbose'}
# default['rackspace_memcached']['config']['-k'] = { value: true, comment: 'lock down all paged memory'}
# default['rackspace_memcached']['config']['-M'] = { value: true, comment: 'return error when memory exhausted' }
# default['rackspace_memcached']['config']['-r'] = { value: true, comment: 'maximize core file limit' }

case node['platform']
when 'redhat', 'centos'
  default['rackspace_memcached']['config']['-u']['value'] = 'memcached'
when 'ubuntu'
  default['rackspace_memcached']['config']['-u']['value'] = 'memcache'
else
  default['rackspace_memcached']['config']['-u']['value'] = 'nobody'
end

default['rackspace_memcached']['include_dev_lib'] = true
